protocols/gossipsub/src/behaviour.rs

```diff
@@ -0,0 +1,1176 @@
+        self.peer_topics.insert(id.clone(), SmallVec::new());
+    }
+
+    fn inject_disconnected(&mut self, id: &PeerId, _: ConnectedPoint) {
+        // TODO: Refactor
+        // remove from mesh, topic_peers, peer_topic and fanout
+        debug!("Peer disconnected: {:?}", id);
+        {
+            let topics = match self.peer_topics.get(&id) {
+                Some(topics) => (topics),
+                None => {
+                    error!("ERROR: Disconnected node, not in connected nodes");
+                    return;
+                }
+            };
+
+            // remove peer from all mappings
+            for topic in topics {
+                // check the mesh for the topic
+                if let Some(mesh_peers) = self.mesh.get_mut(&topic) {
+                    // check if the peer is in the mesh and remove it
+                    if let Some(pos) = mesh_peers.iter().position(|p| p == id) {
+                        mesh_peers.remove(pos);
+                        //TODO: untagPeer
+                    }
+                }
+
+                // remove from topic_peers
+                if let Some(peer_list) = self.topic_peers.get_mut(&topic) {
+                    if let Some(pos) = peer_list.iter().position(|p| p == id) {
+                        peer_list.remove(pos);
+                    //TODO: untagPeer
+                    }
+                    // debugging purposes
+                    else {
+                        warn!(
+                            "ERROR: Disconnected node: {:?} not in topic_peers peer list",
+                            &id
+                        );
+                    }
+                } else {
+                    warn!(
+                        "ERROR: Disconnected node: {:?} with topic: {:?} not in topic_peers",
+                        &id, &topic
+                    );
+                }
+
+                // remove from fanout
+                if let Some(fanout_peers) = self.fanout.get_mut(&topic) {
+                    if let Some(pos) = fanout_peers.iter().position(|p| p == id) {
+                        fanout_peers.remove(pos);
+                    }
+                }
+            }
+        }
+
+        // remove peer from peer_topics
+        let was_in = self.peer_topics.remove(id);
+        debug_assert!(was_in.is_some());
+    }
+
+    fn inject_node_event(&mut self, propagation_source: PeerId, event: GossipsubRpc) {
+        // Handle subscriptions
+        // Update connected peers topics
+        self.handle_received_subscriptions(&event.subscriptions, &propagation_source);
+
+        // Handle messages
+        for message in event.messages {
+            self.handle_received_message(message, &propagation_source);
+        }
+
+        // Handle control messages
+        // group some control messages, this minimises SendEvents (code is simplified to handle each event at a time however)
+        // TODO: Decide if the grouping is necessary
```

Comment:
Given that `IHAVE` messages are per topic and given that messages can have two different topics, in the case a node sends us two `IHAVE` control messages each containing the same message id, without this filtering, this would result in two `IWANT` messages, correct? If so, I think this grouping sounds reasonable and does not introduce a lot of noise.

Reply:
Cool, removed the TODO's